src/tests/payments.js

```diff
@@ -0,0 +1,277 @@
+        handler({
+            origin: this.origin,
+            data: {
+                // Indicate that the user payment flow is
+                // finished. Next fxpay will verify the transaction.
+                status: 'ok',
+            }
+        });
+    };
+    this.removeEventListener = function() {};
+}
+
+
+function FakeProduct() {
+    this.slug = 'some-slug';
+    this.payment_required = true;
+}
+
+
+function purchaseOptions(opt) {
+    opt = opt || {};
+    return {
+        fxpaySettings: {
+            mozApps: null,
+            mozPay: null,
+            window: opt.fakeWindow || new FakeFireplaceWindow(),
+            // On Firefox there is some difference in fxpay's scope
+            // that causes fxpay.init() not to clear this. Hmm.
+            initError: null,
+        }
+    };
+}
+
+
+test('payments: complete payment successfully', function(done, fail) {
+    mock(
+        'payments',
+        fakeDeps(),
+        function(payments) {
+            var product = new FakeProduct();
+            payments.purchase(product, purchaseOptions()).done(function() {
+                logger.log('payment.purchase() is done');
+                done();
+            }).fail(function(_, product, reason) {
+                logger.error('payment.purchase() failed:', reason);
+                fail(reason);
+            });
+        },
+        fail
+    );
+});
+
+test('payments: product already purchased', function(done, fail) {
+    var deps = fakeDeps();
+
+    // Reject the JWT post with a 409 response.
+    deps.requests.promiseToPost = defer.Deferred().reject({}, null, 409).promise();
+
+    mock(
+        'payments',
+        deps,
+        function(payments) {
+            var product = new FakeProduct();
+            payments.purchase(product, purchaseOptions()).done(function() {
+                logger.log('payment.purchase() is done');
+                done();
+            }).fail(function(_, product, reason) {
+                logger.error('payment.purchase() failed:', reason);
+                fail(reason);
+            });
+        },
+        fail
+    );
+});
+
+test('payments: server error', function(done, fail) {
+    var deps = fakeDeps();
+
+    // Reject the JWT post with a server error.
+    deps.requests.promiseToPost = defer.Deferred().reject({}, null, 500).promise();
+
+    mock(
+        'payments',
+        deps,
+        function(payments) {
+            var product = new FakeProduct();
+            payments.purchase(product, purchaseOptions()).done(function() {
+                logger.log('payment.purchase() is done');
+                fail('unexpected success');
+            }).fail(function(_, product, reason) {
+                logger.log('payment.purchase() failed:', reason);
+                eq_(reason, 'MKT_CANCELLED');
+                done();
+            });
+        },
+        fail
+    );
+});
+
+test('payments: product cannot be purchased', function(done, fail) {
```

Comment:
This confused me a bit I was looking for where it was expected to fail. But it's actually more about fxpay.purchase not being called as payment_required is false.